src/middleware/originMiddleware.ts

```diff
@@ -0,0 +1,56 @@
+import { CACHE_HEADERS } from '../constants/cache';
+import type { Context } from '../context';
+import type { Request } from '../routes/request';
+import { isDirectoryPath } from '../utils/path';
+import type { Middleware } from './middleware';
+
+/**
+ * Rewrites request to go to the DO/origin server. This is used as a failover
+ *  for if we exhaust all of our retries to R2.
+ *
+ * This server has the exact same assets as R2. We should minimize the amount
+ *  of requests that hit this middleware unless there's good reason to do so.
+ */
+export class OriginMiddleware implements Middleware {
+  async handle(request: Request, ctx: Context): Promise<Response> {
+    const res = await fetch(
+      `${ctx.env.ORIGIN_HOST}${request.urlObj.pathname}`,
+      {
+        method: request.method,
+        headers: {
+          'user-agent': 'release-cloudflare-worker',
```

Comment:
Empty headers should be undefined when not present, empty string for some of these headers are not valid values for said headers.

Reply:
Although browsers might ignore it, it's not necessarily part of the HTTP spec.

Reply:
Empty headers are omitted from the final response